src/uct/ib/base/ib_md.c

```diff
@@ -98,83 +400,151 @@ static ucs_status_t uct_ib_mem_alloc(uct_md_h uct_md, size_t *length_p, void **a
+    void UCS_V_UNUSED *address = memh->mr->addr;
 
     ucs_memtrack_releasing(&address);
-    return uct_ib_dereg_mr(mr);
+    return uct_ib_memh_dereg(memh);
 }
 
-static ucs_status_t uct_ib_mem_reg(uct_md_h uct_md, void *address, size_t length,
-                                   uct_mem_h *memh_p)
+static ucs_status_t uct_ib_mem_free(uct_md_h md, uct_mem_h memh)
+{
+    uct_ib_memh_t *ib_memh = memh;
+    ucs_status_t status;
+
+    status = uct_ib_mem_free_internal(md, ib_memh);
+    uct_ib_memh_free(ib_memh);
+
+    return status;
+}
+
+static ucs_status_t uct_ib_mem_reg_internal(uct_md_h uct_md, void *address, size_t length,
+                                            uct_ib_memh_t *memh)
 {
     uct_ib_md_t *md = ucs_derived_of(uct_md, uct_ib_md_t);
-    struct ibv_mr *mr;
 
-    mr = ibv_reg_mr(md->pd, address, length, UCT_IB_MEM_ACCESS_FLAGS);
-    if (mr == NULL) {
+    memh->mr = ibv_reg_mr(md->pd, address, length, UCT_IB_MEM_ACCESS_FLAGS);
+    if (memh->mr == NULL) {
         ucs_error("ibv_reg_mr(address=%p, length=%zu, flags=0x%x) failed: %m",
                   address, length, UCT_IB_MEM_ACCESS_FLAGS);
         return UCS_ERR_IO_ERROR;
     }
-
     ucs_trace("registered memory %p..%p on %s lkey 0x%x rkey 0x%x",
-              address, address + length, uct_ib_device_name(&md->dev), mr->lkey,
-              mr->rkey);
+              address, address + length, uct_ib_device_name(&md->dev), memh->mr->lkey,
+              memh->mr->rkey);
+
+    memh->lkey = memh->mr->lkey;
+
+    memh->umr = uct_ib_md_create_umr(md, memh->mr);
 
     UCS_STATS_UPDATE_COUNTER(md->stats, UCT_IB_MD_STAT_MEM_REG, +1);
-    *memh_p = mr;
     return UCS_OK;
 }
 
+static ucs_status_t uct_ib_mem_reg(uct_md_h uct_md, void *address, size_t length,
+                                   uct_mem_h *memh_p)
+{
+    ucs_status_t status;
+    uct_ib_memh_t *memh;
+
+    memh = uct_ib_memh_alloc();
+    if (memh == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    status = uct_ib_mem_reg_internal(uct_md, address, length, memh);
+    if (status != UCS_OK) {
+        uct_ib_memh_free(memh);
+        return status;
+    }
+    *memh_p = memh;
+
+    return UCS_OK;
+}
+
+static ucs_status_t uct_ib_mem_dereg_internal(uct_ib_memh_t *memh)
+{
+    return uct_ib_memh_dereg(memh);
+}
+
 static ucs_status_t uct_ib_mem_dereg(uct_md_h uct_md, uct_mem_h memh)
 {
-    struct ibv_mr *mr = memh;
-    return uct_ib_dereg_mr(mr);
+    ucs_status_t status;
+    uct_ib_memh_t *ib_memh = memh;
+
+    status = uct_ib_mem_dereg_internal(ib_memh);
+    uct_ib_memh_free(ib_memh);
+    return status;
 }
 
 static ucs_status_t uct_ib_mkey_pack(uct_md_h md, uct_mem_h memh,
                                      void *rkey_buffer)
 {
-    struct ibv_mr *mr = memh;
-    *(uint32_t*)rkey_buffer = mr->rkey;
-    ucs_trace("packed rkey: 0x%x", mr->rkey);
+    uct_ib_memh_t *ib_memh = memh;
+    uint32_t *rkeys;
+
+    rkeys    = (uint32_t *)rkey_buffer;
+    rkeys[0] = ib_memh->mr->rkey;
```

Comment:
need to use pack-to-64bit function - now pack is using array and unpack is using >> which is not consistent